### PR TITLE
Refactor sampler

### PR DESCRIPTION
**Pull request by richfitz**
This PR adds basically no new functionality but changes the sampler so that the algorithm is clearer.  With this:

* The comparison function is separate from the sampler, meaning that sampler has nothing to do with the sircovid core model necessarily
* There is a consistent time axis now. The book-keeping for this was horrific enough that I've pulled this into a new function `particle_filter_data` which makes explicit what is going on
* I've updated the code to use a new feature in dde that will speed things up about 20%

This was all done against a small run saved as reference.rds - I've ensured that this agrees but have yet to seriously start unit testing

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -43,4 +43,7 @@ vignettes: vignettes/traduire.Rmd
 	mkdir -p inst/doc
 	cp vignettes/*.html vignettes/*.Rmd inst/doc
 
+inst/extdata/age_distribution.csv: scripts/default_age_distribution.R scripts/WPP2019_INT_F03_1_POPULATION_BY_AGE_ANNUAL_BOTH_SEXES.xlsx
```
Comment (johnlees) [inline Makefile]: Do the xlsx reads work without readxl?

File: R/sampler.R [hunk 2]
```diff
@@ -1,183 +1,270 @@
...
+
+  ## Special treatment for the burn-in phase; later we might use this
+  ## same approach for skipping steps though.
   if (save_particles) {
-    tmp <- mod$run(step = seq(0, (n_days_before_data-1)*time_steps_per_day, time_steps_per_day),
-                 replicate = n_particles)
-    states <- unname(tmp[, seq_along(mod$initial()) + 1L, ])
+    step <- seq(data$step_start[[1L]], data$step_end[[1L]],
+                attr(data, "steps_per_day"))
+    tmp <- model$run(step, use_names = FALSE, replicate = n_particles)
```
Comment (johnlees) [inline R/sampler.R]: More descriptive name than tmp?

File: R/sampler.R [hunk 3]
```diff
@@ -1,183 +1,270 @@
...
+  ## Unpack things that we will use repeatedly
+  phi_ICU <- pars_obs$phi_ICU
+  k_ICU <- pars_obs$k_ICU
+  phi_death <- pars_obs$phi_death
+  k_death <- pars_obs$k_death
+  exp_noise <- pars_obs$exp_noise
+  index_ICU <- c(index$I_ICU) - 1L
+  index_D <- c(index$D) - 1L
+
+  function(t, results, prev_states) {
```
Comment (johnlees) [inline R/sampler.R]: Might be my lack of familiarity with R, but to check:
Is this defining a closure? Might be handy to add a comment to that effect

File: scripts/default_age_distribution.R [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+age_distribution <- function(infile = NULL,
+                             country = "United Kingdom",
+                             year = 2019,
+                             age.limits = c(0, 10, 20, 30, 40, 50, 60, 70, 80)
+                             ) {
+    if (is.null(infile)) {
+      infile <- "scripts/WPP2019_INT_F03_1_POPULATION_BY_AGE_ANNUAL_BOTH_SEXES.xlsx"
+    }
+
+    estimates <- readxl::read_xlsx(infile, sheet = "ESTIMATES", skip = 16)
```
Comment (johnlees) [inline scripts/default_age_distribution.R]: `readxl` dependency or library needed here? Or is this script just run locally to generate data before package upload?

Comment (richfitz) [reply scripts/default_age_distribution.R]: yeah, that's it - run offline, as this takes a while (you'll be seeing this run in your branch I believe, currently caching once per session but that was still annoying me)

File: R/sampler.R (hunk 2 above)
Comment (richfitz) [reply R/sampler.R]: absolutely, sorry I'd hoped to entirely removed that step but have not managed to. I'll update it...

File: Makefile (hunk 1 above)
Comment (richfitz) [reply Makefile]: As noted below - this is just to run offline really

File: R/sampler.R [hunk 5]
```diff
@@ -1,183 +1,281 @@
...
+  ret
 }
 
 
-particle_run_model <- function(pars_model, y, step, mod) {
-  n_particles <- ncol(y)
-  for (k in seq_len(n_particles)) {
-    y[, k] <- mod$run(c(0, step), y[, k],
-                      use_names = FALSE, return_minimal = TRUE)
+##' Compare the model to ICU data for use with the particle filter
```
Comment (johnlees) [inline R/sampler.R:163]: Very minor: inconsistent style with exported function above (which has an extra newline between all of these)

File: tests/testthat/test-sampler.R [hunk 6]
```diff
@@ -3,61 +3,192 @@ context("sampler")
...
   time_steps_per_day <- 4
+  data <- read.csv(sircovid_file("extdata/example.csv"),
+                   stringsAsFactors = FALSE)
+  d <- particle_filter_data(data, "2020-02-02", time_steps_per_day)
 
   pars_model <- generate_parameter_rtm(
-    seed_SSP=10,
-    dt = 1/time_steps_per_day
-  )
+    seed_SSP = 10,
```
Comment (johnlees) [inline tests/testthat/test-sampler.R:13]: This function is removed in #9, but probably best to sort that out after merging both

**Pull request merged**